src/transformers/modeling_utils.py

```diff
@@ -1282,6 +1292,18 @@ def load(module: nn.Module, prefix=""):
 
         return model
 
+    def retrieve_modules_from_weights(self, weight_keys):
+        retrieved_modules = set()
+        for weight_key in weight_keys:
+            module_ptr = self
+            for layer_name in weight_key.split(".")[:-1]:
+                module_ptr = module_ptr._modules[layer_name]
+            retrieved_modules.add(module_ptr)
+        # TOOD: IMPORTANT WE HAVE TO BE SURE THAT ALL WEIGHTS OF THE
+        # RETRIEVED MODULES ACTUALLY HAVE TO BE RANDOMELY INITIALIZED
+        # TEST THAT ALL WEIGHTS OF ALL MODULES ARE IN WEIGHT KEYS
```

Comment:
To make things easier here, we should:
- initialize randomly the modules that have at least one key in the missing keys (as is done here)
- **then** load the state dict
(not sure with the diff in which order this is done right now).

This would still initialize twice some weights in certain edge cases, but only very few of them, and it would be a safer option.

Reply:
Ok, I think I agree here as well - so shis should be the logic then:

1) Load state_dict and set all weights that are loaded as the model weights
2) Randomly initialize all modules that correspond to the missing weights
3) Extract all weights from the state_dict that correspond to the "missing" modules and set them correctly => usually there won't be any weights in the state_dict that corresond to the "missing" modules, but it could be the case in some edge cases

correct?

Reply:
I was thinking of something easier to implement (though your way works too):
- entirely move the logic to determine the missing_keys before the line `load(model_to_load, prefix=start_prefix)` (there is one last part after right now).
- randomly initialize the modules with at least one parameter in the missing keys
- do the `load(model_to_load, prefix=start_prefix)`.

Reply:
Would this also be easily possible if weights are loaded from TF or Flax?

Reply:
In TF we don't add a random init on top of the one TF performs, and for Flax, I think the design already instantiates random weights if we don't pass pretrained ones, but you should know that better than me ;-)

Reply:
What @sgugger said!